### PR TITLE
Check RPIT and AFIT hidden types are well-formed considering regions

### LINKED ISSUE #114728: RPIT hidden types can be ill-formed
The following UB shouldn't compile:
```rust
type Static<'a> = &'static &'a ();
trait Extend<'a> {
    fn extend(self, _: &'a str) -> &'static str;
}
impl<'a> Extend<'a> for Static<'a> {
    fn extend(self, s: &'a str) -> &'static str {
        s
    }
}
fn boom<'a>(arg: Static<'_>) -> impl Extend<'a> {
    arg
}
fn main() {
    let y = boom(&&()).extend(&String::from("temporary"));
    println!("{}", y);
}
```
This regressed in v1.61. I guess it is #94081.

Discovered when investigating #114572.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #114728

r? @aliemjay @lcnr or @oli-obk (who's on vacation but he shall see it when he gets back 🌴)

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 1]
```diff
@@ -470,9 +468,11 @@ fn check_opaque_meets_bounds<'tcx>(
         hir::OpaqueTyOrigin::TyAlias { .. }
             if tcx.def_kind(tcx.parent(def_id.to_def_id())) == DefKind::OpaqueTy => {}
         // Can have different predicates to their defining use
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/check.rs:470]: We could, perhaps, define the assumed wf types of a TAIT to be any types that show up in the bounds of a parent TAIT.

Don't think it's worth unless someone can come up with an unsoundness test here, though.

**Comment by compiler-errors** [comment]
@craterbot check

**Comment by lcnr** [comment]
r? @lcnr
r=me after crater + fcp

@rust-lang/types we previously did not regionck RPITs and async function return types because we assumed that they would already be checked by their parent function.

This is not the case in case we end up "subtyping" into an RPIT, see https://github.com/rust-lang/rust/issues/114728. This PR causes us to now always check regions for RPITs as well.

@rfcbot fcp merge

**Comment by lcnr** [comment]
sorry

@rfcbot cancel

**Comment by lcnr** [comment]
https://github.com/rust-lang/rust/pull/114740#issuecomment-1677961021

@rfcbot fcp merge

**Comment by aliemjay** [comment]
Here is a couple of what I think are unneccessary breakages of this implementation:

- The RPIT variant of #114572:
```rust
struct Static<T: 'static>(T);
fn test<' a>(s: Static<&'static u8>) -> impl Sized + 'a { s } 
```

- Another more contreversial one:
```rust
trait Id<X> { type Ty; }
impl<X> Id<X> for () { type Ty = X; }
fn test() -> impl Id<&'static T, Ty = impl Sized> {}
```

Both are valid patterns that predate #94081. I believe we can keep supporting them by doind the WF check where subtyping is involved, namely in nll_relate and higher-ranked equality (which involves mutual subtyping).

**Comment by lcnr** [comment]
@aliemjay is trying an alternative approach in #114933, still in a crater run. It's definitely worth it to wait until that crater run is finished before making progress here.

**Comment by compiler-errors** [comment]
I'm gonna close this in favor of the other

**Pull request closed without merging**